src/protocol/prf_sharding/feature_label_dot_product.rs

```diff
@@ -0,0 +1,494 @@
+///
+/// Due to limitation in our infra, it's necessary to set the total number of records each channel will ever need to process.
+/// The number of records each channel processes is a function of the distribution of number of records per user.
+/// Rather than calculate this histogram within this function (challenging to do while streaming), at present the caller must pass this in.
+///
+/// The count at a given index indicates the number of users having at least that many rows of data.
+///
+/// Example:
+///   If the input is from 3 users,
+///     - the first having 2 records
+///     - the second having 4 records
+///     - the third having 6 records
+///   Then the histogram that should be provided is:
+///     - [3, 3, 2, 2, 1, 1]
+///
+/// # Errors
+/// Propagates errors from multiplications
+/// # Panics
+/// Propagates errors from multiplications
+#[allow(clippy::async_yields_async)]
+pub async fn compute_feature_label_dot_product<C, FV, F, S>(
+    sh_ctx: C,
+    input_rows: Vec<PrfShardedIpaInputRow<FV>>,
+    histogram: &[usize],
+) -> Result<Vec<S>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + Serializable + SecureMul<C::UpgradedContext<F>>,
+    FV: GaloisField,
+    F: PrimeField + ExtendableField,
+{
+    assert!(FV::BITS > 0);
+
+    // Get the validator and context to use for Gf2 multiplication operations
+    let binary_validator = sh_ctx.narrow(&Step::BinaryValidator).validator::<Gf2>();
+    let binary_m_ctx = binary_validator.context();
+
+    // Get the validator and context to use for `Z_p` operations (modulus conversion)
+    let prime_field_validator = sh_ctx.narrow(&Step::PrimeFieldValidator).validator::<F>();
+    let prime_field_ctx = prime_field_validator.context();
+
+    // Tricky hacks to work around the limitations of our current infrastructure
+    let num_outputs = input_rows.len() - histogram[0];
+    let mut record_id_for_row_depth = vec![0_u32; histogram.len()];
+    let ctx_for_row_number = set_up_contexts(&binary_m_ctx, histogram);
+
+    // Chunk the incoming stream of records into stream of vectors of records with the same PRF
+    let mut input_stream = stream_iter(input_rows);
+    let first_row = input_stream.next().await.unwrap();
+    let rows_chunked_by_user = chunk_rows_by_user(input_stream, first_row);
+
+    // Convert to a stream of async futures that represent the result of executing the per-user circuit
+    let stream_of_per_user_circuits = pin!(rows_chunked_by_user.then(|rows_for_user| {
+        let num_user_rows = rows_for_user.len();
+        let contexts = ctx_for_row_number[..num_user_rows - 1].to_owned();
+        let record_ids = record_id_for_row_depth[..num_user_rows].to_owned();
+
+        for count in record_id_for_row_depth.iter_mut().take(rows_for_user.len()) {
```

Comment:
can be a little shorter and shouldn't cost an extra bounds check because line 262 has the same check

```suggestion
        for count in &mut record_id_for_row_depth[..num_user_rows] {
```